### PR TITLE
Pass `live/4` `:private` values to `socket.private` on mount

### LINKED ISSUE #2136: Private assigns defined on the router not available in socket
I've realized that assigns attached to `private: %{}` from the router are not available on the socket. Is this a bug, intentional or something that currently isn't possible?

My use case is generating dynamic routes from external configuration at compile time, passing an identifier in the route -- as part of the private assigns -- to be able to identify and load that configuration when the route is called. I would like to use hooks to do some pre-checks/authorizations.

Example
```elixir
live "my-path", MyComponent, nil, private: %{page_id: page.id}
```

In `MyComponent`
```elixir
defmodule MyComponent
  use Phoenix, :live_view

  on_mount {__MODULE__, :check_auth}
  
  def on_mount(:check_auth, _params, _session, socket) do
    # socket.private.page_id does not exist
  end
end
```

### PR DESCRIPTION
**Pull request by leandrocp**
An attempt to close #2136

I faced a very similar use case on https://github.com/BeaconCMS/beacon as explained by @barkerja where some live routes are defined at compile-time and having those private values from the router in mount would be really useful.

### CODE REVIEW

File: test/phoenix_live_view/integrations/live_view_test.exs [hunk 1]
```diff
@@ -251,6 +251,13 @@ defmodule Phoenix.LiveView.LiveViewTest do
     end
   end
 
+  describe "private" do
```
Comment (leandrocp) [inline test/phoenix_live_view/integrations/live_view_test.exs:254]: If this approach is good I can add a test for mount hooks to make sure private values are propagated correctly (I haven't tested hooks yet)

**Comment by josevalim** [comment]
I am a bit torn on this one. Today the documentation says it is explicitly for plug. so some people may be putting large data in private, and now it will bloat their session.

Plus, what you put in `private` will stay around forever, unless you clear it, so we need to be careful about automatically injecting things there.

So what if instead we allow a `:session` option. Then you can shove things in the session in a "private" key and store them in private on the `on_mount` callback if/when you need it?

**Comment by leandrocp** [comment]
I get @josevalim concerns and `:session` may give us a better API actually.

We already have `:session` in `live_session/3` and adding `:session` to `live/4` seems like we could merge them together to allow granular session data for particular routes, which would solve the issue. Something like:

```elixir
live_session :default, session: %{foo: 1, bar: 2} do
  live "/status", StatusLive, :index
  live "/feed", FeedLive, :index, session: %{bar: 3, baz: 4}
end
```

```elixir
# /status
def mount(params, session, socket) do
  # session => %{foo: 1, bar: 2}
end

# /feed
def mount(params, session, socket) do
  # session => %{foo: 1, bar: 3, baz: 4}
end

```

Does that make sense?

**Comment by josevalim** [comment]
Can you please send a PR for session?

**Comment by leandrocp** [comment]
> Can you please send a PR for session?

Sure!

**Comment by chrismccord** [comment]
Keep in mind session can be tricky because via live navigate we sign the session to the client and pass it back up. Technically we can merge the route info's session into this decoded session but there's some complexity here where I'm not sure the session is what we want for this given that your data is really not something that needs signed and set back and forth from client to server. Technically you could pass the `metadata` option today and read it back in handle_params via `Phoenix.Router.route_info/4`. I think this is my preference actually and is accommodated by existing features.

**Comment by leandrocp** [comment]
Hey @chrismccord thanks for the feedback!

> pass the metadata option today and read it back in handle_params via Phoenix.Router.route_info/4. I think this is my preference actually and is accommodated by existing features.

I believe it might work for my use case and also for @barkerja so I'm leaving some notes here for future readers and myself:

Given some routes generated at compile-time that we want to carry metadata over to the live view:

```elixir
# router
live_session :default do
  for {path, some_value} <- my_routes do
    live path, PageLive, :index, metadata: %{some_value: some_value}
  end
end
```

```elixir
# live view
defmodule MyApp.PageLive do
  @impl true
  def handle_params(params, uri, socket) do
    %URI{host: host, path: path} = URI.parse(uri)

    case Phoenix.Router.route_info(socket.router, "GET", path, host) do
      %{some_value: some_value} ->
        # do something with `some_value`
        {:noreply, socket}
        
      :error ->
        # deal with the error
        {:noreply, socket}
    end
  end
end
```

**Comment by chrismccord** [comment]
handle_params is part of the mount lifecycle and if you have multiple routes with live navigate that point to the same LV, then you need a place that handles that change (or not) of this sideloaded info so this aligns with defining the special info on the route itself – that special info can change per route, but may be the same LV.

**Comment by chrismccord** [comment]
also if you want to do it once then nothing stops you from doing it once

**Pull request closed without merging**